web/main_js.go

```diff
@@ -18,51 +20,93 @@ var (
 )
 
 func main() {
-	js.Global().Set("GetSnippets", js.FuncOf(GetBlocks))
-	js.Global().Set("GetDocument", js.FuncOf(GetDocument))
-	js.Global().Set("PrepareScript", js.FuncOf(PrepareScript))
+	runme := map[string]interface{}{
+		"initialize":    js.FuncOf(initialize),
+		"getCell":       js.FuncOf(getCell),
+		"getCells":      js.FuncOf(getCells),
+		"getSource":     js.FuncOf(getSource),
+		"updateCell":    js.FuncOf(updateCell),
+		"prepareScript": js.FuncOf(prepareScript),
+	}
+	js.Global().Set("Runme", js.ValueOf(runme))
 
 	select {}
 }
 
-func GetBlocks(this js.Value, args []js.Value) interface{} {
-	readme := args[0].String()
-	blocks := document.NewSource([]byte(readme)).Parse().CodeBlocks()
-
-	var result []interface{}
-
-	for _, block := range blocks {
-		var lines []interface{}
-		for _, line := range block.Lines() {
-			lines = append(lines, line)
-		}
-		entry := map[string]interface{}{
-			"name":        block.Name(),
-			"description": block.Intro(),
-			"content":     block.Content(),
-			"executable":  block.Executable(),
-			"lines":       lines,
-		}
-		result = append(result, entry)
+var (
+	parsed     *document.ParsedSource
+	sourceHash [16]byte
+)
+
+func initialize(this js.Value, args []js.Value) any {
+	source := args[0].String()
+
+	parsed = document.NewSource([]byte(source)).Parse()
+	sourceHash = md5.Sum([]byte(source))
+
+	return nil
+}
+
+func getCell(this js.Value, args []js.Value) any {
+	if parsed == nil {
+		return errors.New("call initialize() first")
+	}
+
+	idx := args[0].Int()
+
+	blocks := parsed.Blocks()
+
+	if idx >= len(blocks) {
+		return errors.New("block index out of range")
 	}
 
+	data, _ := json.Marshal(blocks[idx])
+
+	result := make(map[string]interface{})
+	_ = json.Unmarshal(data, &result)
 	return result
 }
 
-func GetDocument(this js.Value, args []js.Value) interface{} {
-	readme := args[0].String()
-	pSource := document.NewSource([]byte(readme)).Parse()
+func getCells(this js.Value, args []js.Value) any {
+	if parsed == nil {
+		return errors.New("call initialize() first")
+	}
 
 	var b bytes.Buffer
-	_ = renderer.RenderToJSON(&b, pSource.Source(), pSource.Root())
+	_ = renderer.ToJSON(parsed, &b)
+
+	result := make(map[string]interface{})
+	_ = json.Unmarshal(b.Bytes(), &result)
+	return result
+}
+
+func getSource(this js.Value, args []js.Value) any {
+	if parsed == nil {
+		return errors.New("call initialize() first")
+	}
+	return parsed.Source()
```

Comment:
if you try calling this from the browser @adambabik, it'll throw an error. Suspecting it does know how what to do what to do with a byte array.

Reply:
Good catch, I didn't verify in the `example.html` this function. It's fixed and it works now in the browser.